### ISSUE TITLE
Notebook canvas overflows out of the file tab and into the sidebar.

### ISSUE THREAD

**Issue by Anshul7sp1** [microsoft/vscode#129120]
I changed my sidebar location while having a jupyter notebook open. 
It resulted in the notebook canvas overflowing onto the sidebar.
Gets fixed once I switch to some other file then come back.
This glitch appears only when there is a switch of side bar position from left to right.

## Environment data

-   VS Code version: 1.58
-   Jupyter Extension version (available under the Extensions sidebar): v2021.8.1046824664
-   Python Extension version (available under the Extensions sidebar): v2021.7.1050252941
-   OS (Windows | Mac | Linux distro) and version: Windows 11
-   Python and/or Anaconda version: Python 3.8.2 - 64bit
-   Type of virtual environment used (N/A | venv | virtualenv | conda | ...): NA
-   Jupyter server running: Local

## Actual behaviour

Initial state:
![image](https://user-images.githubusercontent.com/52310272/126510249-d6449342-2bea-4e90-8eed-b48fb6a19047.png)

Just after I changed the layout (Move Side Bar Right)
![image](https://user-images.githubusercontent.com/52310272/126510748-3375329a-34b6-4d5d-9eee-fcd2355f9728.png)

## Expected behaviour

After switching to a different file and coming back.

![image](https://user-images.githubusercontent.com/52310272/126511260-0c7b79ff-6c87-4e49-bc9b-ae278a511a3f.png)

## Steps to reproduce:

[**NOTE**: Self-contained, minimal reproducing code samples are **extremely** helpful and will expedite addressing your issue]

1. Create a new empty jupyter notebook
2. Try changing sidebar position. (only left to right switch gives this glitch, the other way is fine)

## Logs

Not sure if this is relevant for this issue

**Comment by abdlrhmanshehatamoussa** [comment]
Hi Guys,

I'm ready to help,

If any one could give a little push on where exactly can find the relevant code and any other recommendations, I'm ready to gooo.

Best of luck !

**Comment by rebornix** [comment]
This is a really good catch.

**Comment by rebornix** [comment]
I can reproduce this issue with both notebook and custom editor. After toggling the position of the sidebar, a `layout` request is triggered after the sidebar moves to the right side, but no event when the activity bar moves to the right side. Thus there is a gap on the left side (whose width is the same as the activity bar)

**Comment by rebornix** [comment]
Apparently `editor.layout` is invoked before setting `left/top` on the editor container DOM node thus the position tracking in notebook and custom editor is measuring position incorrectly. Adding a `setTimeout(() => editor.layout())` can fix the problem but can also be too costly.

**Comment by rebornix** [comment]
A proper fix would be we pass the `left/top` layout info from the grid view to the editor group view.

**Comment by rebornix** [comment]
Fixed in main.

**Comment by bpasero** [comment]
This seems like a general grid layout issue to me unrelated to editors specifically. Thus assigning @joaomoreno and @sbatten (who are on vacation though).

**Comment by mjbvz** [comment]
Both notebooks and normal webviews suffer from this problem. I believe the root cause is the same although the two have different implementations 

After debugging, I believe that the root cause is that we get a call to `layout` before the activity bar has moved to the other side (or before the editor has shifted fully to replace the space the activity bar occupied). Here's a screenshot of the dom state when this happens:

![Screen Shot 2021-08-02 at 12 22 36 PM](https://user-images.githubusercontent.com/12821956/127912684-998b20e9-5d00-4c41-b43c-34ea94b43e51.png)

The important bit is that `left` has been calculated to be `47px` when it should be zero. Here's where the value is calculated:

https://github.com/microsoft/vscode/blob/5c1b0b73d85b3e372690cc18d3b8197643fb5746/src/vs/workbench/contrib/notebook/browser/notebookEditorWidget.ts#L1588

Here you can also see how we are calling layout before main the editor dom note has shifted to the left to replace the activity bar:

![Screen Shot 2021-08-02 at 12 29 44 PM (3)](https://user-images.githubusercontent.com/12821956/127913507-3dab84c6-c174-4528-80c4-2c3e837e3c20.png)



I believe this could be fixed by calling `layout` again once the activity bar (or editor locations) have been fully updated. Ideally we would also avoid calling `layout` until we have the final locations as well. While debugging I noticed that toggling the sidebar already seems to trigger 3 calls to `layout` today

**Comment by bpasero** [comment]
I think the sidebar location change code is here:

https://github.com/microsoft/vscode/blob/9aab65c6ee6fdca6a3e14fbb2efcf52861dd7175/src/vs/workbench/browser/layout.ts#L417-L451

I resort to Steven for a fix who authored this.

**Comment by sbatten** [comment]
After digging into this, I believe https://github.com/microsoft/vscode/commit/0449a18b7bb1d8b3cf7138df63e03126ba47c21d is the reason we hit this issue. @joaomoreno 

Essentially, if a grid gets the same dimensions and position info it will not call layout again on the contained view (in this case the webview or notebook). At the workbench level, after moving the activity bar (the 2nd action in moving the sidebar, we can see that the editor part layout is correctly called.) however, the editor part has several nested grids, one of which does not use top/left info. Since the size and position are the same on the 2nd call, the cache check is hit and we bail before the webview or notebook can get the correct offset info.

It seems this shouldn't work on the first call either but moving the sidebar triggers a size and position change on the top level grid and the activity bar does not.

**Comment by joaomoreno** [comment]
> After digging into this, I believe [0449a18](https://github.com/microsoft/vscode/commit/0449a18b7bb1d8b3cf7138df63e03126ba47c21d) is the reason we hit this issue. @joaomoreno

Does reverting this commit fix this issue?

**Comment by sbatten** [comment]
@joaomoreno yes reverting fixes the issue. I did not, however, because I was not sure of the repercussions.

**Comment by joaomoreno** [comment]
> The hard part of doing the above is making sure positional data gets passed around the composite world, ideally [`Composite.layout`](https://github.com/microsoft/vscode/blob/59f5bbaf253e4e611f814b22a9cac5003eecacbe/src/vs/workbench/browser/composite.ts#L155) would look like this instead:
> 
> ```ts
> 	abstract layout(width: number, height: number, top: number, left: number): void;
> ```
> 
> So maybe @sbatten would give you this. 😉

@sbatten @mjbvz @rebornix Any thoughts on this above?
